crates/bifrost/src/providers/replicated_loglet/tasks/find_tail.rs

```diff
@@ -0,0 +1,645 @@
+            )
+            .await;
+
+            match maybe_info {
+                Ok(Ok(msg)) => {
+                    self.known_global_tail
+                        .notify_offset_update(msg.body().header.known_global_tail);
+                    // We retry on the following errors.
+                    match msg.body().status {
+                        Status::Ok | Status::Sealed => {
+                            return (
+                                self.node_id,
+                                NodeTailStatus::Known {
+                                    local_tail: msg.body().header.local_tail,
+                                    sealed: msg.body().header.sealed,
+                                },
+                            );
+                        }
+                        // retyrable errors
+                        Status::Sealing | Status::Disabled | Status::Dropped => {
+                            // fall-through for retries
+                        }
+                        // unexpected statuses
+                        Status::SequencerMismatch | Status::OutOfBounds | Status::Malformed => {
+                            error!("Unexpected status from log-server node_id={} when getting loglet info for loglet_id={}: {:?}", self.node_id, self.loglet_id, msg.body().status);
+                            return (self.node_id, NodeTailStatus::Unknown);
+                        }
+                    }
+                }
+                Ok(Err(RpcError::SendError(e))) => {
+                    debug!(
+                        "Failed to get loglet info from node_id={} for loglet_id={}: {:?}",
+                        self.node_id, self.loglet_id, e.original
+                    );
+                }
+                Ok(Err(RpcError::Shutdown(_))) => {
+                    // RPC router has shutdown, terminating.
+                    return (self.node_id, NodeTailStatus::Unknown);
+                }
+                Err(_timeout_error) => {
+                    debug!(
+                        "Timeout when getting loglet info from node_id={} for loglet_id={}. Configured timeout={:?} ",
+                        self.node_id, self.loglet_id, request_timeout
+                    );
+                }
+            }
+
+            // Should we retry?
+            if let Some(pause) = retry_iter.next() {
+                trace!(
+                    "Retrying to get loglet info from node_id={} and loglet_id={} after {:?}",
+                    self.node_id,
+                    self.loglet_id,
+                    pause
+                );
+                tokio::time::sleep(pause).await;
+            } else {
+                trace!("Exhausted retries while attempting to get loglet-info from node_id={} and loglet_id={}", self.node_id, self.loglet_id);
+                return (self.node_id, NodeTailStatus::Unknown);
+            }
+        }
+    }
+}
+
+struct WaitForTailOnNode {
```

Comment:
The `FindTailOnNode` and `WaitFotTailOnNode` looks very similar maybe they can become one Generic task, since they only rely mainly on the Header in the incoming message anyway

Reply:
I'd consider this as part of the cosmetic changes that I'll do later.